File: etl/lib/etl/tagging_store.rb
require "csv"
require "fileutils"

module Etl
  class TaggingStore
    TAG_COLUMNS = %w[sleep_issue stress_burnout acupuncture_referral mental_health_referral].freeze

    HEADERS = (%w[pseudonymous_id source_file gender age] + TAG_COLUMNS + %w[personal_report_excerpt]).freeze

    def self.load(path)
      return {} unless File.exist?(path)

      CSV.read(path, headers: true).to_h do |row|
        [row["pseudonymous_id"], row.to_h]
      end
    end

    def self.write(path, rows)
      FileUtils.mkdir_p(File.dirname(path))

      CSV.open(path, "w") do |csv|
        csv << HEADERS
        rows.each do |row|
          csv << HEADERS.map { |h| row[h] }
        end
      end
    end

    def self.blank_tag
      "TODO(Y/N)"
    end
  end
end

File: desktop/etl/taggingStore.js
// Node.js port of ../../etl/lib/etl/tagging_store.rb.
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const { parse } = require("csv-parse/sync");
const { stringify } = require("csv-stringify/sync");

const TAG_COLUMNS = ["sleep_issue", "stress_burnout", "acupuncture_referral", "mental_health_referral"];

const HEADERS = ["pseudonymous_id", "source_file", "gender", "age", ...TAG_COLUMNS, "personal_report_excerpt"];

const BLANK_TAG = "TODO(Y/N)";

function loadTagging(filePath) {
  if (!fs.existsSync(filePath)) return {};

  const content = fs.readFileSync(filePath, "utf8");
  const records = parse(content, { columns: true, skip_empty_lines: true });

  const byId = {};
  for (const row of records) {
    byId[row.pseudonymous_id] = row;
  }
  return byId;
}

function writeTagging(filePath, rows) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });

  const data = rows.map((row) => HEADERS.map((header) => row[header] ?? ""));
  const csv = stringify([HEADERS, ...data]);
  fs.writeFileSync(filePath, csv);
}

module.exports = { TAG_COLUMNS, HEADERS, BLANK_TAG, loadTagging, writeTagging };
